### PR TITLE
Adsk contrib - Fix VC++ compilation warnings & some SonarCloud bugs

### PR DESCRIPTION
**Pull request by hodoulp**
Signed-off-by: Patrick Hodoul <Patrick.Hodoul@autodesk.com>

The pull request mainly changes code around warnings:
- All VC++ warnings directly related to the `pybind11` integration,
- Like the linux platforms, the 'Warnings as errors' in now ON on Window platform for the CI builds,
- All (I hope) new SonarCloud blocker, critical and major bugs.

### CODE REVIEW

File: src/OpenColorIO/fileformats/ctf/CTFReaderHelper.cpp [hunk 1]
```diff
@@ -2564,39 +2564,40 @@ void CTFReaderLogElt_2_0::end()
...
+        getLog()->setBlueParams(bParams);
     }
     else
     {
-        if (!m_baseSet)
+        if (!isBaseSet())
         {
-            if (m_ctfParams.m_style == LogUtil::LOG2 || m_ctfParams.m_style == LogUtil::ANTI_LOG2)
+            if (getCTFParams().m_style == LogUtil::LOG2 
+                    || getCTFParams().m_style == LogUtil::ANTI_LOG2)
```
Comment (michdolan) [inline src/OpenColorIO/fileformats/ctf/CTFReaderHelper.cpp:2595]: [minor] indentation alignment

File: src/OpenColorIO/fileformats/ctf/CTFReaderHelper.cpp [hunk 2]
```diff
@@ -2564,39 +2564,40 @@ void CTFReaderLogElt_2_0::end()
...
             {
                 setBase(2.0);
             }
-            else if (m_ctfParams.m_style == LogUtil::LOG10 ||
-                     m_ctfParams.m_style == LogUtil::ANTI_LOG10)
+            else if (getCTFParams().m_style == LogUtil::LOG10
+                        || getCTFParams().m_style == LogUtil::ANTI_LOG10)
```
Comment (michdolan) [inline src/OpenColorIO/fileformats/ctf/CTFReaderHelper.cpp:2600]: [minor] indentation alignment

File: src/OpenColorIO/fileformats/ctf/CTFReaderHelper.cpp (hunk 1 above)
Comment (hodoulp) [reply src/OpenColorIO/fileformats/ctf/CTFReaderHelper.cpp:2595]: That's how we put multi-line conditions everywhere in the code.

File: src/OpenColorIO/fileformats/ctf/CTFReaderHelper.cpp (hunk 2 above)
Comment (hodoulp) [reply src/OpenColorIO/fileformats/ctf/CTFReaderHelper.cpp:2600]: That's how we put multi-line conditions everywhere in the code.

**Comment by hodoulp** [comment]
Looking at the GitHab Action build, I saw some doc warnings that I fixed here.

**Comment by hodoulp** [comment]
There is something new with the latest MSVC compiler. The `/std:c++11` is not anymore present in the documentation (i.e. only 14, 17, 20 & latest exist). And at the same time, there is a new warning related to a feature of C++17 or C++20 i.e. `nodiscard` on `std::regex_match`. 

To be investigated: When compiling OCIO, what is the C++ version used by MSVC?

**Comment by hodoulp** [comment]
> To be investigated: When compiling OCIO, what is the C++ version used by MSVC?

`/std:c++11` is not supported by recent MSVC versions i.e. only `c++14`, `c++17` and `latest` are currently supported. By default, `latest` is used.

So CMakeLists.txt is now improved:
```
if(MSVC)
	include(CheckCXXCompilerFlag)
	CHECK_CXX_COMPILER_FLAG("-std:c++11" COMPILER_SUPPORTS_CXX11)

	if(NOT COMPILER_SUPPORTS_CXX11 AND ${CMAKE_CXX_STANDARD} EQUAL 11)
		message(STATUS "The compiler ${CMAKE_CXX_COMPILER_ID} has no C++11 only support. Use C++14.")
		set(CMAKE_CXX_STANDARD 14)
	endif()
endif()
```

However, the C++17 warning is still present.

**Comment by hodoulp** [comment]
Still under investigations, please let me merge when ready.

**Comment by hodoulp** [comment]
Post merge comment:
Looking at the Windows STL implementation here (https://github.com/microsoft/STL/blob/ca032aebb01e6a2bd914a8df4b7e77867d1a3e61/stl/inc/regex), some `regex_match` signatures now have the `nodiscard` keyword.

**Pull request merged**